lib/bounces.js

```diff
@@ -10,57 +10,64 @@ const P = require('./promise')
-  const MAX_SOFT = configBounces.soft && configBounces.soft.max || 0
-  const MAX_COMPLAINT = configBounces.complaint && configBounces.complaint.max || 0
-  const DURATION_HARD = configBounces.hard && configBounces.hard.duration || Infinity
-  const DURATION_SOFT = configBounces.soft && configBounces.soft.duration || Infinity
-  const DURATION_COMPLAINT = configBounces.complaint && configBounces.complaint.duration || Infinity
+
   const BOUNCE_TYPE_HARD = 1
   const BOUNCE_TYPE_SOFT = 2
   const BOUNCE_TYPE_COMPLAINT = 3
 
   const freeze = Object.freeze
   const BOUNCE_RULES = freeze({
-    [BOUNCE_TYPE_HARD]: freeze({
-      duration: DURATION_HARD,
-      error: error.emailBouncedHard,
-      max: MAX_HARD
-    }),
-    [BOUNCE_TYPE_COMPLAINT]: freeze({
-      duration: DURATION_COMPLAINT,
-      error: error.emailComplaint,
-      max: MAX_COMPLAINT
-    }),
-    [BOUNCE_TYPE_SOFT]: freeze({
-      duration: DURATION_SOFT,
-      error: error.emailBouncedSoft,
-      max: MAX_SOFT
-    })
+    [BOUNCE_TYPE_HARD]: freeze(configBounces.hard || {}),
+    [BOUNCE_TYPE_SOFT]: freeze(configBounces.soft || {}),
+    [BOUNCE_TYPE_COMPLAINT]: freeze(configBounces.complaint || {})
   })
 
+  const ERRORS = {
+    [BOUNCE_TYPE_HARD]: error.emailBouncedHard,
+    [BOUNCE_TYPE_SOFT]: error.emailBouncedSoft,
+    [BOUNCE_TYPE_COMPLAINT]: error.emailComplaint
+  }
+
   function checkBounces(email) {
     return db.emailBounces(email)
-      .then(bounces => {
-        const counts = {
-          [BOUNCE_TYPE_HARD]: 0,
-          [BOUNCE_TYPE_COMPLAINT]: 0,
-          [BOUNCE_TYPE_SOFT]: 0
+      .then(applyRules)
+  }
+
+  // Relies on the order of the bounces array to be sorted by date,
+  // descending. So, each bounce in the array must be older than the
+  // previous.
+  function applyRules(bounces) {
+    const tallies = {
+      [BOUNCE_TYPE_HARD]: {
+        count: 0,
+        latest: 0
+      },
+      [BOUNCE_TYPE_COMPLAINT]: {
+        count: 0,
+        latest: 0
+      },
+      [BOUNCE_TYPE_SOFT]: {
+        count: 0,
+        latest: 0
+      }
+    }
+    const now = Date.now()
+
+    bounces.forEach(bounce => {
+      const type = bounce.bounceType
+      const ruleSet = BOUNCE_RULES[type]
+      if (ruleSet) {
+        const tally = tallies[type]
+        const tier = ruleSet[tally.count]
+        if (! tally.latest) {
+          tally.latest = bounce.createdAt
         }
-        const now = Date.now()
-        bounces.forEach(bounce => {
-          const type = bounce.bounceType
-          const ruleSet = BOUNCE_RULES[type]
-          if (ruleSet) {
-            if (bounce.createdAt > now - ruleSet.duration) {
-              counts[type]++
-              if (counts[type] > ruleSet.max) {
-                throw ruleSet.error()
-              }
-            }
-          }
-        })
-      })
+        if (tier && bounce.createdAt > now - tier) {
+          throw ERRORS[type](tally.latest)
+        }
+        tally.count++
+      }
```

Comment:
I wonder if it's worth trying to exit this loop early if you exceed the maximum time window for any bounce type, so that we don't even look at e.g. bounce records from two years ago.  Then again, maybe we should just avoid reading those from the db in the first place.  Just a thought, no followup necessary :-)